internal/host/plugin/repository_host_catalog.go

```diff
@@ -143,6 +146,279 @@ func (r *Repository) CreateCatalog(ctx context.Context, c *HostCatalog, _ ...Opt
+	if c == nil {
+		return nil, db.NoRowsAffected, db.NoRowsAffected, errors.New(ctx, errors.InvalidParameter, op, "nil HostCatalog")
+	}
+	if c.HostCatalog == nil {
+		return nil, db.NoRowsAffected, db.NoRowsAffected, errors.New(ctx, errors.InvalidParameter, op, "nil embedded HostCatalog")
+	}
+	if c.PublicId == "" {
+		return nil, db.NoRowsAffected, db.NoRowsAffected, errors.New(ctx, errors.InvalidParameter, op, "no public id")
+	}
+	if c.ScopeId == "" {
+		return nil, db.NoRowsAffected, db.NoRowsAffected, errors.New(ctx, errors.InvalidParameter, op, "no scope id")
+	}
+	if len(fieldMask) == 0 {
+		return nil, db.NoRowsAffected, db.NoRowsAffected, errors.New(ctx, errors.EmptyFieldMask, op, "empty field mask")
+	}
+
+	// Quickly replace the passed in HostCatalog with a clone. This
+	// ensures that we don't alter anything in the original passed in
+	// parameters, which we don't do by convention. We will be adding
+	// to this "working set" as we move through the method.
+	c = c.clone()
+
+	// Get the old catalog first. We patch the record first before
+	// sending it to the DB for updating so that we can run on
+	// OnUpdateCatalog. Note that the field masks are still used for
+	// updating.
+	currentCatalog, _, err := r.LookupCatalog(ctx, c.PublicId)
+	if err != nil {
+		return nil, db.NoRowsAffected, db.NoRowsAffected, errors.Wrap(ctx, err, op, errors.WithMsg(fmt.Sprintf("error looking up catalog with id %q", c.PublicId)))
+	}
+
+	if currentCatalog == nil {
+		return nil, db.NoRowsAffected, db.NoRowsAffected, errors.New(ctx, errors.RecordNotFound, op, fmt.Sprintf("catalog with id %q not found", c.PublicId))
+	}
+
+	// Assert the version of the current catalog to make sure we're
+	// updating the correct one.
+	if currentCatalog.GetVersion() != version {
+		return nil, db.NoRowsAffected, db.NoRowsAffected, errors.New(ctx, errors.VersionMismatch, op, fmt.Sprintf("catalog version mismatch, want=%d, got=%d", currentCatalog.GetVersion(), version))
+	}
+
+	// Clone the catalog so that we can set fields.
+	newCatalog := currentCatalog.clone()
+	var updateAttributes bool
+	var dbMask, nullFields []string
+	for _, f := range fieldMask {
+		switch {
+		case strings.EqualFold("name", f) && c.Name == "":
+			nullFields = append(nullFields, "name")
+			newCatalog.Name = c.Name
+		case strings.EqualFold("name", f) && c.Name != "":
+			dbMask = append(dbMask, "name")
+			newCatalog.Name = c.Name
+		case strings.EqualFold("description", f) && c.Description == "":
+			nullFields = append(nullFields, "description")
+			newCatalog.Description = c.Description
+		case strings.EqualFold("description", f) && c.Description != "":
+			dbMask = append(dbMask, "description")
+			newCatalog.Description = c.Description
+		case strings.EqualFold("attributes", strings.Split(f, ".")[0]):
+			// Flag attributes for updating. While multiple masks may be
+			// sent, we only need to do this once.
+			updateAttributes = true
+		case strings.EqualFold("secrets", f):
```

Comment:
Don't rely on the secret field being present.  If the passed in catalog has anything in the secrets field, just pass that on to the plugin.

Reply:
So this goes back to the question: there's a guard that prevents an empty field mask from being sent in. Do we now get rid of that guard if we are to pass in secrets indiscriminately? If we do that, we lose some ability to validate what has been passed in:

* Fields that have well-defined masks will still be iterated over and accepted properly, or errored out on.
*  `secrets`, or any deeper path thereof, will fail due to invalid field mask.
* Secrets could be passed in without an accompanying field mask. 

It seems to me that this breaks the integrity of the validation work that's done here in processing masks. I know that we discussed adding secrets to the passThroughPrefix to allow for these to be passed in a similar way to attributes, has your thinking on this changed?

Reply:
Just discussed this offline and I think Todd is aware of this and was about to comment on it. Going to proceed with partial validation in the same way that we are doing with attributes.

Reply:
Updated now!